src/applications/mhv-secure-messaging/components/shared/SmRouteNavigationGuard.jsx

```diff
@@ -0,0 +1,93 @@
+import React, { useEffect, useState, useCallback } from 'react';
+import { Prompt } from 'react-router-dom';
+import PropTypes from 'prop-types';
+import { VaModal } from '@department-of-veterans-affairs/component-library/dist/react-bindings';
+
+const SmRouteNavigationGuard = ({
+  when,
+  onConfirmNavigation,
+  modalTitle,
+  modalText,
+  confirmButtonText,
+  cancelButtonText,
+}) => {
+  const [isBlocking, setIsBlocking] = useState(false);
+  const [modalVisible, updateModalVisible] = useState(false);
+
+  const closeModal = () => {
+    setIsBlocking(true); // Keep blocking if the user closes modal
+    updateModalVisible(false);
+  };
+
+  const handleCancel = () => {
+    setIsBlocking(true); // Keep blocking if the user cancels
+    updateModalVisible(false);
+  };
+
+  const handleConfirm = useCallback(
+    e => {
+      onConfirmNavigation(e, true); // true will force navigation and saving, and skip modal
+    },
+    [onConfirmNavigation],
+  );
+
+  const handleBlockedNavigation = useCallback(
+    () => {
+      if (isBlocking) {
+        updateModalVisible(true);
+        setIsBlocking(false);
+        return false; // Blocks Navigation
+      }
+      return true; // Allows Navigation
+    },
+    [isBlocking, setIsBlocking, updateModalVisible],
+  );
+
+  // Update blocking state based on the `when` prop
+  useEffect(
+    () => {
+      setIsBlocking(when);
+    },
+    [when],
+  );
+
+  return (
+    <>
+      <Prompt when={when} message={handleBlockedNavigation} />
+      <VaModal
+        modalTitle={modalTitle}
+        modalText={modalText}
+        onCloseEvent={closeModal}
+        status="warning"
+        visible={modalVisible}
+        data-dd-action-name="Navigation Warning Close Modal Button"
```

Comment:
we may want to check with Anne Costello on these DD tags. They resemble the ones on Compose navigate away. If she does use these for some tracking, we may want to change it to something more distinct may be or at least give her heads up. Not a blocker for this PR, we can tweak it in the future if needed.